src/adapter/src/catalog/open/builtin_item_migration.rs

```diff
@@ -0,0 +1,527 @@
+            PersistClient::CONTROLLER_CRITICAL_SINCE,
+            diagnostics.clone(),
+        )
+        .await
+        .expect("invalid usage");
+    let (mut write_handle, mut read_handle): (
+        WriteHandle<TableKey, ShardId, Timestamp, Diff>,
+        ReadHandle<TableKey, ShardId, Timestamp, Diff>,
+    ) = persist_client
+        .open(
+            shard_id,
+            Arc::new(TableKeySchema),
+            Arc::new(ShardIdSchema),
+            diagnostics,
+            USE_CRITICAL_SINCE_CATALOG.get(persist_client.dyncfgs()),
+        )
+        .await
+        .expect("invalid usage");
+    // Commit an empty write at the minimum timestamp so the shard is always readable.
+    const EMPTY_UPDATES: &[((TableKey, ShardId), Timestamp, Diff)] = &[];
+    let _ = write_handle
+        .compare_and_append(
+            EMPTY_UPDATES,
+            Antichain::from_elem(Timestamp::minimum()),
+            Antichain::from_elem(Timestamp::minimum().step_forward()),
+        )
+        .await
+        .expect("invalid usage");
+
+    // 2. Get the `GlobalId` of all migrated tables.
+    let migrated_tables: BTreeSet<_> = migrated_builtins
+        .into_iter()
+        .filter_map(|id| {
+            if state.get_entry(&id).item().is_storage_collection() {
+                match id {
+                    GlobalId::System(id) => Some(id),
+                    _ => unreachable!("builtin objects must have system ID, found: {id:?}"),
+                }
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // 3. Read in the current contents of the migration shard.
+    let upper = fetch_upper(&mut write_handle).await;
+    let as_of = Antichain::from_elem(upper.saturating_sub(1));
+    let since = read_handle.since();
+    as_of.join(since);
+    let snapshot = read_handle
+        .snapshot_and_fetch(as_of)
+        .await
+        .expect("we have advanced the as_of by the since");
+    soft_assert_or_log!(
+        snapshot.iter().all(|(_, _, diff)| *diff == 1),
+        "snapshot_and_fetch guarantees a consolidated result: {snapshot:?}"
+    );
+    let mut global_id_shards: BTreeMap<_, _> = snapshot
+        .into_iter()
+        .map(|((key, value), _ts, _diff)| {
+            let table_key = key.expect("persist decoding error");
+            let shard_id = value.expect("persist decoding error");
+            (table_key, shard_id)
+        })
+        .collect();
+
+    // 4. Clean up contents of migration shard.
+    let mut migrated_shard_updates: Vec<((TableKey, ShardId), Timestamp, Diff)> = Vec::new();
+    let mut migration_shards_to_finalize = BTreeSet::new();
+    let storage_collection_metadata = {
+        let txn: &mut dyn StorageTxn<Timestamp> = txn;
+        txn.get_collection_metadata()
+    };
+    for (table_key, shard_id) in global_id_shards.clone() {
+        if !migrated_tables.contains(&table_key.global_id)
+            || table_key.deploy_generation != deploy_generation
```

Comment:
I think you want to restrict your attention here to deploy generations that are _less_ than `deploy_generation`. In read only mode, you might even want to halt if you see any shards with a newer generation, as that indicates that _another_ even newer deployment is happening.

Reply:
Good point, just updated this.